File: scripts/leetcode/221maximalSquare.py
from typing import List

class Solution:
    def maximalSquare(self, matrix: List[List[str]]) -> int:
        r = len(matrix)
        c = len(matrix[0])
        max_side = 0
        dp = [[0] * (c + 1) for _ in range(r + 1)]
        for i in range(1, r + 1):
            for j in range(1, c + 1):
                if matrix[i - 1][j - 1] == '1':
                    dp[i][j] = min(dp[i - 1][j], dp[i][j - 1], dp[i - 1][j - 1]) + 1
                    max_side = max(max_side, dp[i][j])
        
        return max_side ** 2


matrix = [['1','0','1','0','0'],['1','0','1','1','1'],['1','1','1','1','1'],['1','0','0','1','0']]
su = Solution()
print(su.maximalSquare(matrix))

File: src/leetcode/221maximalSquare.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

class Solution
{
public:
    int maximalSquare(vector<vector<char>> &matrix)
    {
        int r = matrix.size();
        int c = matrix[0].size();
        vector<vector<int>> dp(r + 1, vector<int>(c + 1));
        int max_side = 0;
        for (int i = 1; i < r + 1; ++i){
            for (int j = 1; j < c + 1; ++j){
                if(matrix[i - 1][j - 1] == '1'){  // 注意C++中单字符用' ',字符串用' '
                    dp[i][j] = min(dp[i - 1][j], min(dp[i][j - 1], dp[i - 1][j - 1])) + 1;
                    max_side = max(max_side, dp[i][j]);
                }
            }
        }
        return max_side * max_side;
    }
};

int main()
{
    vector<vector<char>> matrix = {{'1', '0', '1', '0', '0'}, {'1', '0', '1', '1', '1'}, {'1', '1', '1', '1', '1'}, {'1', '0', '0', '1', '0'}};
    Solution s;
    cout << s.maximalSquare(matrix) << endl;
    return 0;
}
